build-system/tasks/visual-diff.js

```diff
@@ -35,10 +100,463 @@ function setPercyBranch() {
 }
 
 /**
- * Simple wrapper around the ruby based visual diff tests.
+ * Launches a background AMP webserver for unminified js using gulp.
+ *
+ * Waits until the server is up and reachable, and ties its lifecycle to this
+ * process's lifecycle.
+ *
+ * @return {!Promise} a Promise that resolves when the web server is launched
+ *     and reachable.
  */
-function visualDiff() {
-  setPercyBranch();
+async function launchWebServer() {
+  const gulpServeAsync = execScriptAsync(
+      `gulp serve --host ${HOST} --port ${PORT} ${process.env.WEBSERVER_QUIET}`,
+      {stdio: 'inherit'});
+
+  gulpServeAsync.on('exit', code => {
+    if (code != 0) {
+      log('error', colors.cyan("'serve'"),
+          `errored with code ${code}. Cannot continue with visual diff tests`);
+      process.exit(code);
+    }
+  });
+
+  process.on('exit', async() => {
+    if (gulpServeAsync.exitCode == null) {
+      gulpServeAsync.kill();
+      // The child node process has an asynchronous stdout. See #10409.
+      await sleep(100);
+    }
+  });
+
+  let resolver, rejecter;
+  const deferred = new Promise((resolverIn, rejecterIn) => {
+    resolver = resolverIn;
+    rejecter = rejecterIn;
+  });
+  tryConnect({
+    host: HOST,
+    port: PORT,
+    retries: WEBSERVER_TIMEOUT_RETRIES, // retry timeout defaults to 1 sec
+  }).on('connected', resolver).on('timeout', rejecter);
+  return deferred;
+}
+
+/**
+ * Checks the current status of a Percy build.
+ *
+ * @param {string} buildId ID of the ongoing Percy build.
+ * @return {!JsonObject} The full response from the build status server.
+ */
+async function getBuildStatus(buildId) {
+  const statusUri = `${BUILD_STATUS_URL}?build_id=${buildId}`;
+  const {body} = await request(statusUri, {json: true}).catch(error => {
+    log('fatal', 'Failed to query Percy build status:', error);
+  });
+  return body;
+}
+
+/**
+ * Waits for Percy to finish processing a build.
+ * @param {string} buildId ID of the ongoing Percy build.
+ * @return {!JsonObject} The eventual status of the Percy build.
+ */
+async function waitForBuildCompletion(buildId) {
+  log('info', 'Waiting for Percy build', colors.cyan(buildId),
+      'to be processed...');
+  const startTime = Date.now();
+  let lastProgressUpdateTime = Date.now();
+  let status;
+  do {
+    status = await getBuildStatus(buildId);
+
+    if (Date.now() - lastProgressUpdateTime >=
+            BUILD_PROCESSING_PROGRESS_REPORT_MS) {
+      log('info', 'Still waiting for Percy build', colors.cyan(buildId),
+          'to be processed...');
+      lastProgressUpdateTime = Date.now();
```

Comment:
I wonder how useful this is. It used to be necessary back when we had really long processing times. Consider removing / simplifying at a later point.